src/native/minipal/utf8.c

```diff
@@ -0,0 +1,2131 @@
+                if (InRange(ch, HIGH_SURROGATE_START, LOW_SURROGATE_END))
+                {
+                    // 4 byte encoding - high surrogate + low surrogate
+
+                    int chd = *pSrc;
+                    if (
+                        ch > HIGH_SURROGATE_END ||
+                        !InRange(chd, LOW_SURROGATE_START, LOW_SURROGATE_END))
+                    {
+                        // Back up and drop out to slow loop to figure out error
+                        pSrc--;
+                        break;
+                    }
+                    pSrc++;
+
+                    // byteCount - this byte is compensated by the second surrogate character
+                }
+                byteCount++;
+            }
+            byteCount++;
+
+            // byteCount - the last byte is already included
+        }
+
+        // no pending char at this point
+        ch = 0;
+    }
+
+#if WIN64
+    // check for overflow
+    assert(byteCount >= 0);
+#endif
+    assert(!fallbackUsed || self->buffer.encoder.fallbackCount < 0);
+
+    return byteCount;
+}
+
+size_t minipal_get_length_utf8_to_utf16(const char* source, size_t sourceLength, unsigned int flags)
+{
+    errno = 0;
+
+    if (sourceLength == 0)
+        return 0;
+
+    UTF8Encoding enc =
+    {
+        .buffer = { .decoder = { .fallbackCount = -1, .fallbackIndex = -1, .strDefault = { 0xFFFD, 0 }, .strDefaultLength = 1 } },
+        .useFallback = !(flags & MINIPAL_MB_NO_REPLACE_INVALID_CHARS),
+#if BIGENDIAN
+        .treatAsLE = (flags & MINIPAL_TREAT_AS_LITTLE_ENDIAN)
+#endif
+    };
+
+    return GetCharCount(&enc, (unsigned char*)source, sourceLength);
+}
+
+size_t minipal_get_length_utf16_to_utf8(const CHAR16_T* source, size_t sourceLength, unsigned int flags)
+{
+    errno = 0;
+
+    if (sourceLength == 0)
+        return 0;
+
+    UTF8Encoding enc =
+    {
+        // repeat replacement char (0xFFFD) twice for a surrogate pair
+        .buffer = { .encoder = { .fallbackCount = -1, .fallbackIndex = -1, .strDefault = { 0xFFFD, 0xFFFD, 0 }, .strDefaultLength = 2 } },
+        .useFallback = true,
+#if BIGENDIAN
+        .treatAsLE = (flags & MINIPAL_TREAT_AS_LITTLE_ENDIAN)
+#endif
+    };
+
+#if !BIGENDIAN
+    (void)flags; // unused
+#endif
+
+    return GetByteCount(&enc, (CHAR16_T*)source, sourceLength);
+}
+
+size_t minipal_convert_utf8_to_utf16(const char* source, size_t sourceLength, CHAR16_T* destination, size_t destinationLength, unsigned int flags)
+{
+    size_t ret;
+    errno = 0;
+
+    if (sourceLength == 0)
+        return 0;
+
+    UTF8Encoding enc =
+    {
+        .buffer = { .decoder = { .fallbackCount = -1, .fallbackIndex = -1, .strDefault = { 0xFFFD, 0 }, .strDefaultLength = 1 } },
+        .useFallback = !(flags & MINIPAL_MB_NO_REPLACE_INVALID_CHARS),
+#if BIGENDIAN
+        .treatAsLE = (flags & MINIPAL_TREAT_AS_LITTLE_ENDIAN)
+#endif
+    };
+
+    if (GetCharCount(&enc, (unsigned char*)source, sourceLength) > destinationLength)
```

Comment:
In `minipal_convert_*` APIs, ideally we should do the re-count upfront. But the current (existing) behavior is that we do not overwrite the memory in case of insufficient buffer. Not sure if it matters in this error case; it shouldn't, but I kept it as-is to preserve the original behavior.

Reply:
I do not think we need to preserve the original behavior in this case. It should be fine to partially overwrite the output in this case.

Reply:
There will be additional 20 branches in `GetChars()` and `GetBytes()` (combined) if we remove this condition. I suppose that will still perform better than upfront counting.

Reply:
Right. It is also better from security point of view (e.g. consider situation where the input changes between counting and actual encoding due to some unrelated bug).